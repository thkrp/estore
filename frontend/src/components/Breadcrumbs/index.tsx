import React, { useEffect, useState } from 'react';
import { ReactComponent as LoaderIcon } from '../../assets/icons/horizontal-loader.svg';
import {
    CategoriesStyled,
    ElementStyled,
    LinkStyled,
    ListStyled,
    MoreCategoriesButtonStyled,
    WrapperStyled
} from './index.styles';
import useBreadcrumbs from '../../common/hooks/use.breadcrumbs';
import { transformToKey } from '../../common/helpers/key.helper';

const Breadcrumbs = () => {
    const breadcrumbs = useBreadcrumbs();
    const [showCategories, setShowCategories] = useState<Record<string, boolean>>({});

    useEffect(() => {
        setShowCategories({});
    }, [breadcrumbs]);

    return (
        <WrapperStyled>
            <ListStyled role="navigation" aria-label="breadcrumbs">
                {breadcrumbs.length
                    ? breadcrumbs.map((el, index) =>
                          index === breadcrumbs.length - 1 ? (
                              <ElementStyled key={transformToKey(index, el.code)}>
                                  {el.name || <LoaderIcon height="15px" role="img" />}
                              </ElementStyled>
                          ) : (
                              <ElementStyled key={transformToKey(index, el.code)}>
                                  <LinkStyled to={el.url}>{el.name || el.code}</LinkStyled>
                                  {el.children && (
                                      <>
                                          {breadcrumbs.length - 1 !== index && (
                                              <MoreCategoriesButtonStyled
                                                  role="presentation"
                                                  onClick={() =>
                                                      setShowCategories(prev => ({
                                                          [el.code]: !prev[el.code]
                                                      }))
                                                  }
                                              />
                                          )}
                                          {showCategories[el.code] && (
                                              <CategoriesStyled role="menu" aria-label="sub-categories">
                                                  {el.children.map((cat, i) => (
                                                      <LinkStyled
                                                          to={cat.url}
                                                          key={transformToKey(i, `${cat.url}`)}
                                                          role="menuitem"
                                                      >
                                                          {cat.name || cat.code}
                                                      </LinkStyled>
                                                  ))}
                                              </CategoriesStyled>
                                          )}
                                      </>
                                  )}
                              </ElementStyled>
                          )
                      )
                    : null}
            </ListStyled>
        </WrapperStyled>
    );
};

export default Breadcrumbs;
