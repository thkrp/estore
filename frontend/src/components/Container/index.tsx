import React, { FC, ReactNode } from 'react';
import { WrapperStyled } from './index.styles';

type Props = {
    children: ReactNode;
    className?: string;
};
/**
 * A wrapper for content that limits the width.
 */
const Container: FC<Props> = ({ children, className }) => (
    <WrapperStyled className={className}>{children}</WrapperStyled>
);

export default Container;
