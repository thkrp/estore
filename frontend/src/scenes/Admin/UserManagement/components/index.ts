import ActionBar from './ActionBar';
import UserList from './UserList';

export { UserList, ActionBar };
