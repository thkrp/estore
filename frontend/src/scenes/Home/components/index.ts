import BestSales from './BestSales';
import Tabs from './Tabs';

export { Tabs, BestSales };
