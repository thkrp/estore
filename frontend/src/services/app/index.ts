import {
    AppResponse,
    Localization,
    BottomMenu,
    CatalogMenu,
    TopMenu,
    GeneralInfo,
    TranslationHashResponse,
    TranslationResponse
} from 'app-shared';
import { axiosAdapter } from '../../common/helpers/axios-adatpter';

class AppService {
    async fetchAppInfo(): Promise<AppResponse<GeneralInfo>> {
        const { data } = await axiosAdapter.doGet<AppResponse<GeneralInfo>>({ url: '/v1/general-information' });
        return data;
    }

    async fetchTopMenu(): Promise<AppResponse<TopMenu>> {
        const { data } = await axiosAdapter.doGet<AppResponse<TopMenu>>({ url: '/v1/navigation/top-menu' });
        return data;
    }

    async fetchBottomMenu(): Promise<AppResponse<BottomMenu>> {
        const { data } = await axiosAdapter.doGet<AppResponse<BottomMenu>>({ url: '/v1/navigation/bottom-menu' });
        return data;
    }

    async fetchCatalogMenu(): Promise<AppResponse<CatalogMenu>> {
        const { data } = await axiosAdapter.doGet<AppResponse<CatalogMenu>>({ url: '/v1/navigation/catalog-menu' });
        return data;
    }

    async getTranslation(localization: Localization): Promise<AppResponse<TranslationResponse>> {
        const { data } = await axiosAdapter.doGet<AppResponse<TranslationResponse>>({
            url: `/v1/translation/${localization}`
        });
        return data;
    }

    async getTranslationHash(localization: Localization): Promise<AppResponse<TranslationHashResponse>> {
        const { data } = await axiosAdapter.doGet<AppResponse<TranslationHashResponse>>({
            url: `/v1/translation/hash/${localization}`
        });

        return data;
    }
}

export const appService = new AppService();
