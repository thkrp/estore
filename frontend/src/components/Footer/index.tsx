import React, { FC, ReactElement } from 'react';
import { BottomMenu, BottomMenuItemDto } from 'app-shared';
import {
    FooterStyled,
    ListHeadingStyled,
    ListStyled,
    ListItemStyled,
    ListWrapperStyled,
    LogoDescriptionStyled,
    LogoWrapperStyled,
    MenuWrapperStyled,
    WrapperStyled
} from './index.styles';
import Container from '../Container';
import { FooterInfo } from '../../common/types/footer.info';
import Logo from '../Logo';
import Link from '../Links/Link';
import ContactLink from '../Links/ContactLink';
import ExternalLink from '../Links/ExternalLink';

type Props = {
    info?: FooterInfo;
    menu?: BottomMenu[];
};

enum LinkTypes {
    section = 'SECTION',
    link = 'LINK',
    email = 'EMAIL',
    phone = 'PHONE',
    external = 'EXTERNAL'
}

const Links: { [key in LinkTypes]: (item: BottomMenuItemDto) => ReactElement } = {
    [LinkTypes.section]: ({ url, name }) => <Link to={url}>{name}</Link>,
    [LinkTypes.link]: ({ url, name }) => <Link to={url}>{name}</Link>,
    [LinkTypes.phone]: ({ name }) => <ContactLink value={name} type="phone" hiddenIcon />,
    [LinkTypes.email]: ({ name }) => <ContactLink value={name} type="email" hiddenIcon />,
    [LinkTypes.external]: ({ url, name }) => <ExternalLink href={url} title={name} />
};

const Footer: FC<Props> = ({ menu, info }) => {
    return (
        <WrapperStyled>
            <Container>
                <FooterStyled>
                    <LogoWrapperStyled>
                        <Logo logo={info?.logo_footer} />
                        <LogoDescriptionStyled>{info?.footer_description}</LogoDescriptionStyled>
                    </LogoWrapperStyled>
                    <MenuWrapperStyled>
                        {!!menu?.length &&
                            menu.map(list => (
                                <ListWrapperStyled key={list.name}>
                                    <ListHeadingStyled>{list.name}</ListHeadingStyled>
                                    <ListStyled>
                                        {list.items.map(li => {
                                            const LinkComponent = Links[li.type as LinkTypes];
                                            return <ListItemStyled key={li.name}>{LinkComponent(li)}</ListItemStyled>;
                                        })}
                                    </ListStyled>
                                </ListWrapperStyled>
                            ))}
                    </MenuWrapperStyled>
                </FooterStyled>
            </Container>
        </WrapperStyled>
    );
};

export default Footer;
