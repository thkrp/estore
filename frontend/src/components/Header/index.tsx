import React, { FC } from 'react';
import { GeneralInfo, TopMenu } from 'app-shared';
import TopNav from '../Navigation/Top';
import { HeaderStyled } from './index.styles';
import Container from '../Container';
import Logo from '../Logo';
import Actions from './Actions';
import Cart from './Cart';

type Props = {
    menu?: TopMenu;
    info?: GeneralInfo;
};

const Header: FC<Props> = ({ menu, info }) => {
    return (
        <header>
            <TopNav menu={menu} />
            <Container>
                <HeaderStyled>
                    <Logo logo={info && info.logo} />
                    <Actions phones={info?.phones} emails={info?.emails} />
                    <Cart />
                </HeaderStyled>
            </Container>
        </header>
    );
};

export default Header;
