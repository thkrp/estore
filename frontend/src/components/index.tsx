import Container from './Container';
import Header from './Header';
import Footer from './Footer';
import CatalogNav from './Navigation/Catalog';
import TopNav from './Navigation/Top';
import Breadcrumbs from './Breadcrumbs';
import Link from './Links/Link';
import HorizontalSlider from './Sliders/Horizontal';
import BrandCard from './Cards/BrandCard';
import ProductCard from './Cards/ProductCard';
import SpecialCategories from './SpecialCategories';
import HomeSlider from './Sliders/Home';
import InputText from './InputText';
import MobileNav from './Navigation/Mobile';
import Search from './Search';
import CleanHtml from './Html';
import Login from './MaterialUi/Auth/Login';

export {
    Container,
    Header,
    Footer,
    CatalogNav,
    TopNav,
    Breadcrumbs,
    Link,
    HorizontalSlider,
    ProductCard,
    BrandCard,
    SpecialCategories,
    HomeSlider,
    InputText,
    MobileNav,
    Search,
    CleanHtml,
    Login
};
