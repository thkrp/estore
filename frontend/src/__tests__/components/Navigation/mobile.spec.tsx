/* eslint-disable no-restricted-imports*/
import React from 'react';
import '@testing-library/jest-dom';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';
import { Locale, Localization, mockData } from 'app-shared';
import { MemoryRouter } from 'react-router-dom';
import * as sanitizer from '@braintree/sanitize-url';
import { IntlProvider } from 'react-intl';
import { MobileNav } from '../../../components';
import { Menu } from '../../../store/app/types/app.state';

describe('MobileNav', () => {
    afterEach(cleanup);
    const menu: Partial<Menu> = { top: mockData.top };

    it('should be defined', () => {
        const { container } = render(
            <IntlProvider key={Localization.ru} locale={Locale.ru} messages={mockData.localizationData}>
                <MobileNav menu={menu} info={mockData.generalInformation} />
            </IntlProvider>,
            { wrapper: MemoryRouter }
        );
        expect(container).toBeDefined();
        expect(container).toMatchSnapshot();
    });

    it('sanitizeUrl has been called', () => {
        jest.spyOn(sanitizer, 'sanitizeUrl');
        render(
            <IntlProvider key={Localization.ru} locale={Locale.ru} messages={mockData.localizationData}>
                <MobileNav menu={menu} />
            </IntlProvider>,
            { wrapper: MemoryRouter }
        );
        expect(sanitizer.sanitizeUrl).toHaveBeenCalled();
    });

    it('more button should open and close additional menu', () => {
        render(
            <IntlProvider key={Localization.ru} locale={Locale.ru} messages={mockData.localizationData}>
                <MobileNav menu={menu} />
            </IntlProvider>,
            { wrapper: MemoryRouter }
        );
        const moreButton = screen.getByRole('button');
        expect(screen.queryAllByRole('navigation').length).toBe(1);
        fireEvent.click(moreButton);
        expect(screen.queryAllByRole('navigation').length).toBe(2);
        fireEvent.click(moreButton);
        expect(screen.queryAllByRole('navigation').length).toBe(1);
    });
});
