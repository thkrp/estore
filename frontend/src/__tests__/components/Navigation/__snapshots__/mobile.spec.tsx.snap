// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogNav should be defined 1`] = `
<div>
  <div
    class="sc-clcPSL ejIKSr"
  >
    <div
      class="sc-fbzYXw cRYaIo"
    >
      <a
        class="sc-GFXod dLNAup"
        href="/"
      >
        Home
      </a>
      <a
        class="sc-GFXod dLNAup"
        href="/catalog"
      >
        Catalog
      </a>
      <a
        class="sc-GFXod dLNAup"
        href="/cart"
      >
        Cart
      </a>
      <button
        class="sc-fXitXF hjVjVE"
        type="button"
      >
        More
      </button>
    </div>
  </div>
</div>
`;
